データベース設計

nana0817 opened the pull request:
# What
ER図とREADMEを作成。

# Why
データベース設計のため。

モデリング
https://i.gyazo.com/e7aa4c8ad8ad13d46551216953bffe4b.png
ER図
https://i.gyazo.com/3b49dd8a744a3f1a4d81109c0a61cf21.png

nana0817 committed:
ER図作成

nana0817 committed:
ER図　カラム名　修正

nana0817 committed:
モデリング・ER図　修正

モデリング
新しく「配送先に関する情報」を追加
それに伴って、「購入に関する情報」の中身・それぞれのアソシエーションを修整

ER図
addressesテーブルを追加
それに伴って、ordersテーブルのカラム・それぞれのアソシエーションを修正

nana0817 committed:
README　作成

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
-Things you may want to cover:
+| Column                        | Type    | Options                   |
+| ----------------------------- | ------- | ------------------------- |
+| nickname                      | string  | null: false, unique: true |
```
```ruby
| nickname                      | string  | null: false, unique: true |
```

ニックネームは重複しても問題ありませんので、一意性制約を外しておきましょう！

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| nickname                      | string  | null: false, unique: true |
+| email                         | string  | null: false, unique: true |
+| password                      | string  | null: false               |
+| password_confirmation         | string  | null: false               |
```
```ruby
| password                      | string  | null: false               |
| password_confirmation         | string  | null: false               |
```
上記のパスワードの情報を保存するカラムの設定ですが、片方を削除し、もう片方のカラム名を`encrypted_password`に変更しておきましょう！

理由：
Gemのdeviseでは`encrypted_password`というカラムがデフォルトで用意されているためです。

ビューで自身が入力するときは`password`と`password_confirmation`というキーを使用しますが、データベースに保存される前に、deviseが2つの情報を照合します。また、一致していればパスワードそのものを暗号化してくれる仕組みになっています。

その暗号化された後のパスワードが`encrypted_password`です。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| Chinese_character_family_name | string  | null: false               |
+| Chinese_character_first_name  | string  | null: false               |
+| Katakana_family_name          | string  | null: false               |
+| Katakana_first_name           | string  | null: false               |
```
```ruby
| Chinese_character_family_name | string  | null: false               |
| Chinese_character_first_name  | string  | null: false               |
| Katakana_family_name          | string  | null: false               |
| Katakana_first_name           | string  | null: false   
```

カラム名はスネークケースが命名規則となりますので、大文字ではなく小文字+アンダーバーで作成しましょう！

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| birth_year                    | integer | null: false               |
+| birth_month                   | integer | null: false               |
+| birth_date                    | integer | null: false               |
 
```
```ruby
| birth_year                    | integer | null: false               |
| birth_month                   | integer | null: false               |
| birth_date                    | integer | null: false               |
```

生年月日を保存するカラムは3つを1つにまとめていただき、型を`date`型に変更しましょう！
date型は年、月、日を1つでまとめて指定して保存できるためです。

今回の場合ですと、年月日を一括で管理することでカラムの準備も最小限にできますので、
`birth_day`等の名前でカラムを１つにして、date型への変更をお願いいたします。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| Column          | Type       | Options                        |
+| --------------- | ---------- | ------------------------------ |
+| title           | string     | null: false                    |
+| text            | text       | null: false                    |
```
```ruby
| text            | text       | null: false                    |
```

textカラムに関しましては、別のカラム名に変更いただきますようお願いいたします！

textという単語は予約語と称されるもので、railsのアプリケーション内ですでに使用されている単語となります。
今回はtext型で既出です。

予約語を使用した場合、アプリ作成中に突然`arugument error`などのエラーが頻出する可能性が高まりますので、お勧めしておりません。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| condition       | string     | null: false                    |
+| shipping_cost   | string     | null: false                    |
+| shipping_source | string     | null: false                    |
+| shipping_days   | string     | null: false                    |
```
```ruby
| category        | string     | null: false                    |
| condition       | string     | null: false                    |
| shipping_cost   | string     | null: false                    |
| shipping_source | string     | null: false                    |
| shipping_days   | string     | null: false                    |
```
商品の状態・配送料の負担などのカラムはactive_hashを用いての実装となりますので、以下を参考に修正をお願いいたします。

active_hashを用いたカラムの特徴は2点ございます：
①カラム名に`_id`が付与されること
active_hashを用いたカラムは、具体的な値ではなくid値（数値）を保存します。
この場合は以下のように、カラムに_idを付与しておくことで管理がしやすくなります。

```ruby
| category_id        | string     | null: false                    |
※単数形_idで作成すること。
```

②integer型を付与すること
id値（数値）を保存する際は、string型ではなく
integer型でカラムを作成すると管理しやすくなります。
以下のように、string型からinteger型への修正をお願いいたします。

```ruby
| category_id        | integer     | null: false                    |

```

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| shipping_source | string     | null: false                    |
+| shipping_days   | string     | null: false                    |
+| price           | integer    | null: false                    |
+| user_id         | references | null: false, foreign_key: true |
```
```ruby
| user_id         | references | null: false, foreign_key: true |
```

references型を使用する際は、カラム名に_idの付与は必要ありません！
migrateしたタイミングで自動的に付与されるためです。

他にも同様の修正が必要な部分がご自身で修正をお願いいたします。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| ------------------- | ---------- | ------------------------------ |
+| user_id             | references | null: false, foreign_key: true |
+| item_id             | references | null: false, foreign_key: true |
+| address_id          | references | null: false, foreign_key: true |
```
```ruby
| address_id          | references | null: false, foreign_key: true |
```

ordersテーブルは購入履歴のみを保存したいので、上記は削除しておきましょう！

その代わり、ordersテーブルとaddressesテーブルの結びつきは、addressesテーブル側で行います。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| building_name    | string     |                                |
+| telephone_number | integer    | null: false                    |
+| user_id          | references | null: false, foreign_key: true |
+
```
```ruby
| user_id          | references | null: false, foreign_key: true |
```

ユーザー情報はordersテーブルですでに保存できていますので重複して保存する必要はありません！

その代わり、ordersテーブルの外部キーカラムを用意しましょう。
双方の結びつきを明確にするために必要となります。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認いただけますと幸いです。
また、コメントの修正に伴ってアソシエーションも修正が出てくるかと存じますが、
その場合はご自身で適宜修正いただいて問題ございません。

修正後、再度レビュー依頼フォームのご提出ください。
また、次回ご提出いただく際は最新のER図を頂戴してもよろしいでしょうか？

ご対応のほど、よろしくお願いいたします。

nana0817 commented:
コードレビューありがとうございます！
修正にとりかかります！

nana0817 committed:
usersテーブルnicknameカラムのOptions修正

今回の場合、ニックネームは重複しても問題ないので、一意性制約を外した。

nana0817 committed:
usersテーブルpassword_confirmationカラムをencrypted_passwordに変更

これから導入するGemのdeviseにはencrypted_passwordカラムがデフォルトで備わっているため。

ビューで自身が入力するときはpasswordとpassword_confirmationというキーを使用するが、データベースに保存される前に、deviseが2つの情報を照合する。また、一致していればパスワードそのものを暗号化してくれる仕組みになっている。
その暗号化された後のパスワードがencrypted_password。

nana0817 committed:
usersテーブル　カラム名の頭文字を小文字に修正

カラム名はスネークケースが命名規則となっているので、小文字＋アンダーバーの形に変更。

nana0817 committed:
usersテーブル　生年月日を表すカラム名およびカラムの型を変更

生年月日を表すカラムを一つにまとめ、型をdate型に変更。

nana0817 committed:
itemsテーブル　「商品の説明」を表すカラムのカラム名変更

カラム名をtextからdetailsに変更
textは予約語であり、アプリ作成中にarugument errorなどのエラーが頻出する可能性が高まるので、カラム名を変更。

nana0817 committed:
itemsテーブル　「商品の詳細」「配送について」に関するカラムのカラム名とカラムの型の変更

商品の状態・配送料の負担などのカラムはactive_hashを用いての実装となるので、
①カラム名は単数形_idの形で
②カラムの型はinteger型で
の条件に合うように変更。

nana0817 committed:
外部キーのカラムのカラム名の変更

references型を使用する場合は、カラム名に「_id」を付ける必要はない。
「_id」はmigrateしたタイミングで自動的に付与される。

nana0817 committed:
orders, addressesテーブルの外部キーカラム変更

今回のアプリの場合、ordersテーブルには購入履歴のみ保存したいので、外部キーカラム「address」は削除。
代わりに、addressesテーブルに外部キーカラム「order」を追加。

nana0817 committed:
addressesテーブル　外部キーカラム「user」削除

ユーザー情報はordersテーブルですでに保存できており、重複して保存する必要はないため、削除。

nana0817 committed:
README アソシエーションを修正

nana0817 committed:
ER図の修正

nana0817 commented:
ご指摘いただいた点は全て修正しました！

以下が最新のER図になります。
https://i.gyazo.com/e4cb00285ce21e99ba2dfc8c66251c98.png

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| -------------------------------- | ------- | ------------------------- |
+| nickname                         | string  | null: false               |
+| email                            | string  | null: false, unique: true |
+| password                         | string  | null: false               |
```
◯カラムを削除しましょう！

| password | string | null: false |

encrypted_passwordにて登録されているため

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,68 @@
 
-* System dependencies
+- has_many :items
+- has_one :order
```
◯アソシエーションを修正しましょう！

購入履歴は複数保持が可能です

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| Column                | Type        |  Options                       |
+| --------------------- | ----------- | ------------------------------ |
+| title                 | string      | null: false                    |
+| details               | text        | null: false                    |
```
◯カラム名を修正しましょう！

単数表記に修正

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| Column           | Type       | Options                        |
+| ---------------- | ---------- | ------------------------------ |
+| post_code        | integer    | null: false                    |
+| prefecture       | string     | null: false                    |
```
◎都道府県に関するカラムを修正しましょう！

```
| prefecture | string | null: false |
```

都道府県カラムにつきましては、商品テーブル・配送先テーブルのカラム名を統一する必要がございます。

都道府県(prefecture_id)等のカラム名にて２つのテーブルで内容を揃えていきましょう！

nana0817 replied:
itemsテーブルは送り元、addressesテーブルは送り先を表すものですが、
カラム名を統一したほうがいいんですか？

ShotaOtake commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| municipalities   | string     | null: false                    |
+| street_address   | string     | null: false                    |
+| building_name    | string     |                                |
+| telephone_number | integer    | null: false                    |
```
◎電話番号カラムの型を文字列型に修正してきましょう！
→integer型だと0から始まる数値を扱うことが出来ないため

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※DB設計にて修正箇所のございます際は、お手数ですが、最新のER図のキャプチャ画像も改めて添付頂きますようお願い致します。
```

nana0817 commented:
コードレビューありがとうございます！
修正に取り掛かります！

nana0817 committed:
usersテーブル　passwordカラムを削除

新規登録の際にパスワードは「encrypted_password」カラムに保存されるので、「password」カラムを他に用意する必要はない。

nana0817 committed:
users, ordersテーブルのアソシエーション　修正

1人のユーザーは購入履歴をを複数所持することができるので、１対多のhas_manyに変更

nana0817 committed:
itemsテーブル　detailsカラムのカラム名を単数表記に変更

カラム名は単数表記にするのが命名規則なので、「item_description」に変更。
加えて、titleカラムもわかりやすいように、「item_name」に変更。

nana0817 committed:
items, addressesテーブル　都道府県カラムのカラム名を統一

今回のアプリの場合、都道府県を表すカラムの名前は二つのテーブルで統一しなければならないので、変更。
それに伴って、addressesテーブルの都道府県カラムのカラムの型もinteger型に変更。

nana0817 committed:
addressesテーブル　電話番号カラムのカラムの型を変更

integer型だと、0から始まる数値を扱うことができないため。

nana0817 committed:
README　スペース調整

nana0817 committed:
モデリング・ER図の変更

nana0817 commented:
ご指摘いただいた点は全て修正しました！

以下が最新のER図です。
https://i.gyazo.com/5609d9dfd1004ad986fb84d40ae3566a.png

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,67 @@
+
+| Column           | Type       | Options                        |
+| ---------------- | ---------- | ------------------------------ |
+| post_code        | integer    | null: false                    |
```
```Rb
| post_code | integer | null: false |
```
郵便番号はintegerではなくstringで実装しましょう。
integerの場合`123-4567`などが引き算で計算されてしまい、`-4444`となり正しく保存できないためです。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしました。
ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

nana0817 committed:
addresesテーブル　post_codeカラムのカラムの型を変更

integer型のままだと、引き算として計算されてしまい、うまく保存されないため、string型に変更。
例えば、123-4567が-4444となってしまう。

nana0817 commented:
ご指摘いただいた点を全て修正しました！

最新のER図です。
https://i.gyazo.com/a2e59cce2fc34c168d68000334ccbf06.png

mihot117 commented:
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
